### PR TITLE
feat: Extra lrp to avoid off ramp

### PR DESCRIPTION
**Pull request by BonyBabu**

### CODE REVIEW

File: encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import java.util.stream.StreamSupport;
+
+public class OfframpProcessor extends AbstractLrpProcessor {
+
+    private static double BEARING_RANGE_OF_SIBLING_LINES = 45;
+
+    public static class RouteNodeInfo {
+        public Node node;
+        public double lengthViaNode;
+        public RouteNodeInfo(Node node, double lengthViaSNode) {
```
Comment (stephencurran) [inline encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java]: is the `S` in the `lengthViaSNode` a typo?

File: encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    private OfframpProcessor(OpenLREncoderProperties properties) {
+        this.properties = properties;
+    }
+
+    public static OfframpProcessor with(OpenLREncoderProperties properties) {
+        return new OfframpProcessor(properties);
+    }
+
```
Comment (stephencurran) [inline encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java]: Here is an unnecessary new line.  You just need one new line between methods.

File: encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        List<Integer> intermediates = new ArrayList<>();
+
+        if (route.isEmpty()) {
+            return intermediates;
+        }
+
+        Line lrpLine = route.get(0);
+        double bearingOfLrpLine = GeometryUtils.calculateLineBearing(lrpLine, GeometryUtils.BearingDirection.IN_DIRECTION, this.properties.getBearingDistance(), 0);
+        Node startNode = route.get(0).getStartNode();
+        List<Line> siblingsOfLrpLine = StreamSupport.stream(Spliterators.spliteratorUnknownSize(startNode.getOutgoingLines(), 0), false).filter(line -> line != lrpLine).collect(Collectors.toList());
```
Comment (stephencurran) [inline encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java]: Such chained operations is best to do over multiple lines.  It makes it more readable.

```java
List<Line> siblingsOfLrpLine = StreamSupport.stream(Spliterators.spliteratorUnknownSize(startNode.getOutgoingLines(), 0), false)
    .filter(line -> line != lrpLine)
    .collect(Collectors.toList());
```

File: encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        }
+
+        GeoCoordinates source = startNode.getGeoCoordinates();
+        GeoCoordinates destination = route.get(route.size() - 1).getEndNode().getGeoCoordinates();
+        Map<Integer, Double> distanceViaNodes = new HashMap<>();
+
+        for (int index = 1; index < route.size(); ++index) {
+            Line line = route.get(index);
+            RouteNodeInfo relevantNodeInfo = getRelevantRouteNode(source, line, destination);
+            distanceViaNodes.put(index, relevantNodeInfo.lengthViaNode);
```
Comment (stephencurran) [inline encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java]: Do we need the `RouteNodeInfo` class?  We return an instance of it via the `getRelevantRouteNode` but then don't seem to use the `Node` property of it.  Can we instead just return this airline distance property from the method instead?  Then we can call the method `getAirlineDistanceViaNode` and can remove the `RouteNodeInfo` class.

File: encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        }
+
+        Integer intermediatePoint = distanceViaNodes.entrySet().stream().sorted(Collections.reverseOrder(Map.Entry.comparingByValue()))
```
Comment (stephencurran) [inline encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java]: When chaining operations like this, one operation per line makes it much more readable.

```java
Integer intermediatePoint = distanceViaNodes.entrySet().stream()
    .sorted(Collections.reverseOrder(Map.Entry.comparingByValue()))
    .filter(cp -> cp.getValue() != 0)
    .map(cp -> cp.getKey())
    .findFirst()
    .orElse(null);
```

File: encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+    /**
+     * @return true if postprocessor is not configured in properties file;
+     */
+    protected final boolean isNotActive() {
+        return !properties.insertLrpToAvoidOffRamp();
+    }
+
+    public List<Integer> determineNewIntermediatePoints(List<Line> route) throws OpenLRProcessingException {
+
```
Comment (stephencurran) [inline encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java]: You can get rid of this new line here.

File: encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+import openlr.map.GeoCoordinates;
+import openlr.map.Line;
+import openlr.map.Node;
+import openlr.map.utils.GeometryUtils;
+
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+public class OfframpProcessor extends AbstractLrpProcessor {
```
[the remark this answers is not in the document]
Comment (stephencurran) [reply encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java]: I wonder whether `OffRamp` is the best name for this.  This applies to all cases where the road forks by an angle of less than 45 degrees, not just in the case of off ramps.  Would `RoadForkProcessor` be a better name?

File: encoder/src/main/java/openlr/encoder/postprocessor/AbstractLrpProcessor.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        int from = lrpPositions.get(lrpPositions.size() - 1);
+        LocRefPoint lastPoint = new LocRefPoint(route.subList(from, route.size()), properties);
+        revisedLrpList.add(lastPoint);
+        return revisedLrpList;
+    }
+
+    /**
+     * @return true if postprocessor is not configured in properties file;
+     */
+    protected abstract boolean isNotActive();
```
[the remark this answers is not in the document]
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/postprocessor/AbstractLrpProcessor.java]: f6c41f62659ef9375da2b7438f0985e9587cb108

File: encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+    private static double BEARING_RANGE_OF_SIBLING_LINES = 45;
```
[the remark this answers is not in the document]
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java]: f6c41f62659ef9375da2b7438f0985e9587cb108

File: encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java]: f6c41f62659ef9375da2b7438f0985e9587cb108

File: encoder/src/main/java/openlr/encoder/postprocessor/AbstractLrpProcessor.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    protected abstract List<Integer> determineNewIntermediatePoints(List<Line> route) throws OpenLRProcessingException;
+
+    /**
+     * @param route route between the adjacent location reference points(including the road segment of the destination lrp)
+     * @param lrpPositions indices of the road segment where to insert the intermediate points.
+     * @return list of location reference points with intermediate points
+     * @throws OpenLRProcessingException
+     */
+    private List<LocRefPoint> createNewLRPs(List<Line> route, List<Integer> lrpPositions) throws OpenLRProcessingException {
+        List<LocRefPoint> revisedLrpList = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/postprocessor/AbstractLrpProcessor.java]: f6c41f62659ef9375da2b7438f0985e9587cb108

File: encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java (hunk 3 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java]: f6c41f62659ef9375da2b7438f0985e9587cb108

File: encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java (hunk 4 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java]: f6c41f62659ef9375da2b7438f0985e9587cb108

File: encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java (hunk 5 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java]: f6c41f62659ef9375da2b7438f0985e9587cb108

File: encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java (hunk 6 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java]: f6c41f62659ef9375da2b7438f0985e9587cb108

File: encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java (hunk 1 above)
Comment (BonyBabu) [reply encoder/src/main/java/openlr/encoder/postprocessor/OfframpProcessor.java]: f6c41f62659ef9375da2b7438f0985e9587cb108

**Pull request closed without merging**